beacon_node/http_api/src/lib.rs

```diff
@@ -2434,9 +2434,17 @@ pub fn serve<T: BeaconChainTypes>(
                     let execution_optimistic =
                         chain.is_optimistic_or_invalid_head().unwrap_or_default();
 
+                    let state_slot = (epoch + 1).end_slot(T::EthSpec::slots_per_epoch());
+                    let state_root = chain.state_root_at_slot(state_slot).unwrap().unwrap();
+                    let finalized = chain
+                        .is_finalized_state(&state_root, state_slot)
+                        .unwrap_or_default();
```

Comment:
I think it's best to include this logic in the `beacon_chain.compute_attestation_rewards` method for the consistency, as the state used for computing the response is derived in that method. 

I've noticed that the `execution_optimistic` above uses the value from the `head` block instead of the state used for computation, which I think is incorrect, so would be nice to fixed it here too.

That could mean changing the method to return a tuple `Result<(StandardAttestationRewards, execution_optimistic, finalized), BeaconChainError>`.

Reply:
Also please avoid using `.unwrap()` here as it could potentially cause panic.

Reply:
Checking the state root is unnecessary IMO. We should just check:

```
let finalized = epoch <= finalized_checkpoint.epoch;
```

Where `finalized_checkpoint` is read from `chain.canonical_head.cached_head().finalized_checkpoint()`. This is less complicated and also faster because it avoids reading from disk.

Reply:
Nice!

Reply:
Why are we checking for epoch + 2 here?

Reply:
The state we read is at the end of epoch N + 1, and it becomes finalized when the block at the start of N + 2 is finalized.

It's maybe a bit open to interpretation whether we should report the finalization of the epoch itself, or the finalization of the epoch rewards. I think the latter is more useful.

Reply:
So, this flag is basically for epoch rewards finalizationm which take 2 epochs extra to process.

Reply:
Yeah